akka-http-core/src/main/scala/akka/http/impl/engine/http2/Http2Demux.scala

```diff
@@ -332,6 +337,11 @@ private[http2] abstract class Http2Demux(http2Settings: Http2CommonSettings, ini
           pullFrameIn()
         }
 
+        override def onUpstreamFinish(): Unit = {
+          onTryComplete()
```

Comment:
Do we need this? If we get completion from the connection/TLS it's not likely that we will be able to send anything else. So maybe better to keep it simple.

Reply:
As discussed on the4 call, I think we can/have_to keep this.

Reply:
That's not quite what I meant :)

Without this the new tests in Http2ServerSpec are failing, yes. But I don't think we should change the behavior for incoming completion of the network for now. As I said before half-closed TLS/HTTP/2 connections are not really supported. In any case, the peer would have to send a GOAWAY first before closing the stream.

When I commented about server behavior on a previous review I meant something else, i.e. a potential issue on the server side if the user side closes the response side of the handler which will now go into the special shutdown state with no opportunity to actually complete potentially open streams which are still waiting for a response.

Reply:
I will look into those issue to get this PR done.